source/compiler/qsc_frontend/src/resolve.rs

```diff
@@ -1638,52 +1234,90 @@ impl GlobalTable {
 
             let namespace = self
                 .scope
-                .insert_or_find_namespace_from_root(global_namespace, root);
+                .insert_or_find_namespace(global_namespace.clone(), package_root);
 
             match (global.kind, global.visibility) {
                 (global::Kind::Ty(ty), hir::Visibility::Public) => {
+                    self.scope
+                        .terms
+                        .get_mut_or_default(namespace)
+                        .insert(global.name.clone(), Res::Item(ty.id, global.status));
                     self.scope
                         .tys
                         .get_mut_or_default(namespace)
-                        .insert(global.name, Res::Item(ty.id, global.status));
+                        .insert(global.name.clone(), Res::Item(ty.id, global.status));
+                    self.scope.importables.get_mut_or_default(namespace).insert(
+                        global.name,
+                        Res::Importable(Importable::Ty(ty.id, global.status)),
+                    );
                 }
-                (global::Kind::Term(term), visibility) => {
+                (global::Kind::Callable(term), visibility) => {
                     if visibility == hir::Visibility::Public {
                         self.scope
                             .terms
                             .get_mut_or_default(namespace)
                             .insert(global.name.clone(), Res::Item(term.id, global.status));
+
+                        self.scope.importables.get_mut_or_default(namespace).insert(
+                            global.name.clone(),
+                            Res::Importable(Importable::Callable(term.id, global.status)),
+                        );
                     }
                     if term.intrinsic {
                         self.scope.intrinsics.insert(global.name);
                     }
                 }
-                (global::Kind::Namespace, hir::Visibility::Public) => {
-                    self.scope.insert_or_find_namespace(global.namespace);
+                (global::Kind::Namespace(item_id), hir::Visibility::Public) => {
+                    let this_namespace = self
+                        .scope
+                        .insert_or_find_namespace(vec![global.name.clone()], Some(namespace));
+
+                    self.scope.importables.get_mut_or_default(namespace).insert(
+                        global.name.clone(),
```

Comment:
This clone is technically not needed:

```suggestion
                        global.name,
```